drivers/i3c/i3c_npcx.c

```diff
@@ -0,0 +1,2052 @@
+				if (target != NULL) {
+					target->dynamic_addr = 0;
+				}
+
+				break;
+			}
+		}
+
+		/* Check all targets have been assigned DA and DAA complete */
+	} while ((!IS_BIT_SET(inst->MSTATUS, NPCX_I3C_MSTATUS_COMPLETE)) &&
+		 npcx_i3c_state_get(inst) != MSTATUS_STATE_IDLE);
+
+out_do_daa:
+	/* Exit DAA mode when error occurs */
+	if (ret != 0) {
+		npcx_i3c_request_emit_stop(inst);
+	}
+
+	/* Clear all flags. */
+	npcx_i3c_errwarn_clear_all(inst);
+	npcx_i3c_status_clear_all(inst);
+
+	/* Re-Enable I3C IRQ sources. */
+	npcx_i3c_interrupt_enable(inst, intmask);
+
+	npcx_i3c_fifo_flush(inst);
+	npcx_i3c_mutex_unlock(dev);
+
+	return ret;
+}
+
+/*
+ * brief:  Send Common Command Code (CCC).
+ *
+ * param[in] dev      Pointer to controller device driver instance.
+ * param[in] payload  Pointer to CCC payload.
+ *
+ * return:  The same as i3c_do_ccc()
+ *          0 If successful.
+ *          -EBUSY Bus is busy.
+ *          -EIO General Input / output error.
+ *          -EINVAL Invalid valid set in the payload structure.
+ *          -ENOSYS Not implemented.
+ */
+static int npcx_i3c_do_ccc(const struct device *dev, struct i3c_ccc_payload *payload)
+{
+	const struct npcx_i3c_config *config = dev->config;
+	int ret;
+	struct i3c_reg *inst = config->base;
+	uint32_t intmask;
+	int xfered_len;
+
+	if (dev == NULL || payload == NULL) {
+		return -EINVAL;
+	}
+
+	npcx_i3c_mutex_lock(dev);
+
+	/* Disable interrupt */
+	intmask = inst->MINTSET;
+	npcx_i3c_interrupt_all_disable(inst);
+
+	/* Clear status and flush fifo */
+	npcx_i3c_xfer_reset(inst);
+
+	LOG_DBG("CCC[0x%02x]", payload->ccc.id);
+
+	/* Write emit START and broadcast address (0x7E) */
+	ret = npcx_i3c_request_emit_start(inst, I3C_BROADCAST_ADDR, NPCX_I3C_MCTRL_TYPE_I3C, false,
+					  0);
+	if (ret < 0) {
+		LOG_ERR("CCC[0x%02x] %s START error (%d)", payload->ccc.id,
+			i3c_ccc_is_payload_broadcast(payload) ? "broadcast" : "direct", ret);
+
+		goto out_do_ccc;
+	}
+
+	/* Write CCC command */
+	npcx_i3c_status_clear_all(inst);
+	npcx_i3c_errwarn_clear_all(inst);
+	xfered_len = npcx_i3c_xfer_write_fifo(inst, &payload->ccc.id, 1, payload->ccc.data_len > 0);
+	if (xfered_len < 0) {
+		LOG_ERR("CCC[0x%02x] %s command error (%d)", payload->ccc.id,
+			i3c_ccc_is_payload_broadcast(payload) ? "broadcast" : "direct", ret);
+		ret = xfered_len;
+
+		goto out_do_ccc;
+	}
+
+	/* Write data (defining byte or data bytes) for CCC if needed */
+	if (payload->ccc.data_len > 0) {
+		npcx_i3c_status_clear_all(inst);
+		npcx_i3c_errwarn_clear_all(inst);
+		xfered_len = npcx_i3c_xfer_write_fifo(inst, payload->ccc.data,
+						      payload->ccc.data_len, false);
+		if (xfered_len < 0) {
+			LOG_ERR("CCC[0x%02x] %s command payload error (%d)", payload->ccc.id,
+				i3c_ccc_is_payload_broadcast(payload) ? "broadcast" : "direct",
+				ret);
+			ret = xfered_len;
+
+			goto out_do_ccc;
+		}
+
+		/* Write back the transferred bytes */
+		payload->ccc.num_xfer = xfered_len;
+	}
```

Comment:
How does this distinguish the difference between the defining byte and payload for the direct CCC in I3C v1.1?

Reply:
Only the i3c_ccc APIs will know the difference (e.g., i3c_ccc_do_getstatus()). 
This part handles both broadcast and direct CCC write data operations in the earlier stage(marked as red in the picture below).

![image](https://github.com/zephyrproject-rtos/zephyr/assets/150219805/7fd21301-98d6-4f0f-a63b-8f28fe121bbf)